### DISCUSSION TITLE
MLX90614 Console "mlx Checksum Error"

### DISCUSSION THREAD

**Discussion by ghost** [arendst/Tasmota#19903] [Q&A]
EDIT: I Use the Addafruit MLX90614 2.1.3 Library and the MLXTEST Sketch
Wrote this in MLX90614 Console "mlx Checksum Error" #15821
Hello, I have the same problem and I think it is the Library and not the wiring and not the Sensor itself. 
I say this because I also use the the Arduino Test sketch without any error. The sketch uses the hardware I2C wires of ESP8266 
D1SCL  D2SDA  So I gave TASMOTA a second try with this pins, but no luck. 
I may have to do with the link speed or some other critical timings but it is not the sensor or connectors.
Arduino Sketch Serial Monitor give me:
20:10:03.309 -> Ambient = 79.75*F	Object = 80.08*F
20:10:03.309 -> 
20:10:03.781 -> Ambient = 26.53*C	Object = 26.71*C
20:10:03.781 -> Ambient = 79.75*F	Object = 80.08*F
20:10:03.781 -> 
20:10:04.313 -> Ambient = 26.53*C	Object = 26.85*C
20:10:04.313 -> Ambient = 79.75*F	Object = 80.33*F
20:10:04.313 -> 
20:10:04.809 -> Ambient = 26.53*C	Object = 26.91*C
20:10:04.809 -> Ambient = 79.75*F	Object = 80.44*F

With Tasmota I get:
20:07:09.433 mlx checksum error
20:07:10.438 mlx checksum error
20:07:11.416 mlx checksum error
20:07:12.423 mlx checksum error
20:07:13.451 mlx checksum error
20:07:14.418 mlx checksum error
20:07:15.427 mlx checksum error
20:07:16.444 mlx checksum error
20:07:17.449 mlx checksum error
20:07:18.411 mlx checksum error
20:07:19.435 mlx checksum error
20:07:20.441 mlx checksum error
20:07:21.449 mlx checksum error
20:07:22.421 mlx checksum error
20:07:24.432 mlx checksum error
20:07:25.408 mlx checksum error
20:07:27.421 mlx checksum error
20:07:29.455 mlx checksum error
20:07:30.414 mlx checksum error
20:07:31.445 mlx checksum error

_Originally posted by @Solar320 in https://github.com/arendst/Tasmota/discussions/15821#discussioncomment-7449161_

**Comment by ghost** [comment]
![Fix-Screenshot 2023-11-03 at 18-52-30 Tasmota - Main Menu](https://github.com/arendst/Tasmota/assets/111298158/b9668fb0-083b-4ff0-bd62-bb08df855eee)
![Fix-Screenshot 2023-11-03 at 18-53-30 Tasmota - Console](https://github.com/arendst/Tasmota/assets/111298158/6dc849c3-c0a7-4013-861f-3660f0813923)
I changed **this** in the xsns_46_MLX90614.ino and now the MLX is working without CRC Errors.

**I don't know why the Author don't want that this Sensor is working, since this error seems to be there for over 1  year.**

For some reason the Author of this Driver has inserted there two lines and commented out the first line and commented in the second line.
If I turn that arround the sensor is working and deliver an value every 2 seconds without any CRC error.
I don't know yet why he use/test  two variants to read the Temperatures out of the sensor.

Since I am not real know what I am doing, may be someone with more experience should look over this error.
I use VisualStudioCode for compiling some code and change some things there and there but I am far away from real programming.
For now I try to work with this Workarround and hope I can use the IR Temp with my Thermostat, because this was the Goal of this Projekt.

void MLX90614_Every_Second(void)
{
    mlx90614.i2c_buf = I2cRead24(I2_ADR_IRT, MLX90614_TOBJ1);  **//EIN changed to ON**   <<<<<<<<<<<<<<-----
    //mlx90614.value = MLX90614_read16(I2_ADR_IRT, MLX90614_TOBJ1);  **//AUS changed to OFF** <<<<<<<<<-----
    if (mlx90614.value & 0x8000) {
      mlx90614.obj_temp = -999;
    } else {
      mlx90614.obj_temp = ((float)mlx90614.value * 0.02f) - 273.15f;
    }
    mlx90614.i2c_buf = I2cRead24(I2_ADR_IRT,MLX90614_TA);        **//EIN changend to ON**  <<<<<<<<<<<<<<-----
    //mlx90614.value = MLX90614_read16(I2_ADR_IRT, MLX90614_TA); **//AUS changed to OFF** <<<<<<<<<<<<----
    if (mlx90614.value & 0x8000) {
      mlx90614.amb_temp = -999;
    } else {
      mlx90614.amb_temp = ((float)mlx90614.value * 0.02f) - 273.15f;
    }
}

**Comment by Staars** [comment]
Well, basically you completely reverted this commit:
https://github.com/arendst/Tasmota/commit/803ecbd9a6dd03907bb2fbab375f187fabf2e355

This is the "old" version of the driver now and the added functions of the commit are bypassed. Thus no CRC computation is done and can not fail.

I assume the "reverted" commit was tested successfully back in time and did solve a problem. So it is possible, that different setups behave differently. 
@gemu2015 Do you remember, if you had the MLX90614 connected to an ESP8266 and/or ESP32?

(I will try to find my MLX90614 in some of the drawers ...)

**Reply by gemu2015** [reply]
at that time i tested mlx90614 only on esp8266. i just tested the current driver again on esp8266 and see no errors.
will test on esp32 in the next few days

**Comment by Staars** [comment]
I will try to jump in too. This could be an edge case.

@Solar320 If I read correctly, you use an ESP8266. If yes, can you try an ESP32 too?

**Reply by gemu2015** [reply]
ok, also on ESP32 no errors with the current driver!

**Reply by Staars** [reply]
Thank you for the quick test!

I2cRead24() does basically the same under the hood as the current implementation in the driver and at least for me it does not explain a "better" reading. 
Only real difference in the work around is the bypass of the CRC checksum test, which may fail for @Solar320 for unknown reasons as the reported sensor values seem to make sense.

**Comment by Staars** [comment]
Finally found my sensor and tested it with ESP8266 and ESP32-C3.
Both work with the current driver for me too.

@Solar320 I fear you have a non-optimal hardware setup, as you see the check of a potential  error condition - sadly revealing an error. If you leave out the CRC check (like in the old Tasmota driver or Adafruits library example), then everything seems to work, but in fact it does not. Even consistent wrong values that look plausible are wrong.

**Comment by ghost** [comment]
I found out that it is not primary has to do with the Tasmota code.
I tested 2 pcs of ESP8266. One is an ESP8266-12E and the other one is an ESP8266-MOD.
Both brings up some crc errors from time to time.  I saw that the ESP8266-12E has unexplainable temperatur differences between obj and Amb temp of 3 °C, the ESP8266-MOD will not have this differences <=0.3K. (the sensor in both cases lays down on the Table at the same place to prevent to measure wrong values) I think the -12E has some internal errors, I don't use it anymore, now I use the ESP32 or the -MOD version. At the first time I used another -12E and this thing produce CRC errors on mass. May be someone can reproduce this error at the 12E model or not.  

For me this can be closed, if someone has the same problem there should be warned for the -12E model.
May be it is only my spezial board type which produce this errors, I don't know....

![ESP8266-12E](https://github.com/arendst/Tasmota/assets/111298158/c9d797e3-2e44-4db5-8d1b-4fc4f3017433)
![ESP8266-12E-Tasmota - Main Menu](https://github.com/arendst/Tasmota/assets/111298158/2ec832a9-29e0-4359-93b0-7be6f646c5f3)
![ESP8266MOD DS18b20 plus MLX90614-amthermostat-Tasmota - Main Menu](https://github.com/arendst/Tasmota/assets/111298158/259316f5-fdd7-4561-8e6c-b03bb8929a10)
![ESP8266MOD](https://github.com/arendst/Tasmota/assets/111298158/6d499453-ed2d-44ea-9387-794857e3b0a7)
